---
layout: paper
title: "Evaluation of mitochondrial DNA copy number estimation techniques."
image: /assets/images/Default_paper.png
authors: Ryan J Longchamps, Christina A Castellani, Stephanie Y Yang, Charles E Newcomb, Jason A Sumpter, John Lane, Megan L Grove, Eliseo Guallar, Nathan Pankratz, Kent D Taylor, Jerome I Rotter, Eric Boerwinkle, Dan E Arking
ref: Longchamps et al. 2020. PLoS One.
journal: "PloS one <b>15</b>, e0228166 (2020)"
doi: 10.1371/journal.pone.0228166
abbrev: "PLoS One <b>15</b>, e0228166 (2020)"
pub_year: 2020
---

<br />
<div data-badge-popover="right" data-badge-type="donut" data-pmid="32004343" data-hide-no-mentions="true" class="altmetric-embed"></div>

# Abstract

Mitochondrial DNA copy number (mtDNA-CN), a measure of the number of mitochondrial genomes per cell, is a minimally invasive proxy measure for mitochondrial function and has been associated with several aging-related diseases. Although quantitative real-time PCR (qPCR) is the current gold standard method for measuring mtDNA-CN, mtDNA-CN can also be measured from genotyping microarray probe intensities and DNA sequencing read counts. To conduct a comprehensive examination on the performance of these methods, we use known mtDNA-CN correlates (age, sex, white blood cell count, Duffy locus genotype, incident cardiovascular disease) to evaluate mtDNA-CN calculated from qPCR, two microarray platforms, as well as whole genome (WGS) and whole exome sequence (WES) data across 1,085 participants from the Atherosclerosis Risk in Communities (ARIC) study and 3,489 participants from the Multi-Ethnic Study of Atherosclerosis (MESA). We observe mtDNA-CN derived from WGS data is significantly more associated with known correlates compared to all other methods (p < 0.001). Additionally, mtDNA-CN measured from WGS is on average more significantly associated with traits by 5.6 orders of magnitude and has effect size estimates 5.8 times more extreme than the current gold standard of qPCR. We further investigated the role of DNA extraction method on mtDNA-CN estimate reproducibility and found mtDNA-CN estimated from cell lysate is significantly less variable than traditional phenol-chloroform-isoamyl alcohol (p = 5.44x10-4) and silica-based column selection (p = 2.82x10-7). In conclusion, we recommend the field moves towards more accurate methods for mtDNA-CN, as well as re-analyze trait associations as more WGS data becomes available from larger initiatives such as TOPMed.

